### PR TITLE
[PyTorch] Add max_score support for MuonClip

### PR DESCRIPTION
**Pull request by cyanguwa**
# Description

This PR adds `max_score` support in `FusedAttention` and `UnfusedDotProductAttention` backends in TE-PyTorch. The `max_score` can be used by a [MuonClip](https://arxiv.org/pdf/2502.16982) optimizer. This PR only support FP16/BF16 and non-CP cases.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Add max_score support in FusedAttention and UnfusedDotProductAttention

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/pytorch/attention/test_attention.py [hunk 1]
```diff
@@ -1025,13 +1052,18 @@ def get_dummy_cuda_rng_tracker() -> CudaRNGStatesTracker:
         fast_zero_fill=True,
     )
     if is_training:
-        out.backward(d_out)
+        out.backward((d_out, torch.zeros(1,device="cuda")))
+
+    if config.return_max_score:
+        out = (out, max_score)
+    else:
+        out = (out, None)
```
Comment (skyw) [inline tests/pytorch/attention/test_attention.py]: This changes returned values when max score is not requested. Should we keep backward compatibility if new feature is not used?

Comment (cyanguwa) [reply tests/pytorch/attention/test_attention.py]: This is only in the test. The actual code/implementation only outputs `out, max_score` when `return_max_score=True`.

**Comment by BoxiangW** [comment]
LGTM thx

**Comment by BoxiangW** [comment]
One more thing on this PR, I think we agreed before on changing the naming into `max_logit` or `max_qk_logit` since it represent the value before softmax op

**Comment by vcherepanov-nv** [comment]
LGTM

**Review by vcherepanov-nv — APPROVED** [review]
